File: public/docs/_examples/structural-directives/dart/lib/heavy_loader_component.dart
// #docregion
import 'dart:async';

import 'package:angular2/core.dart';

int nextId = 1;

@Component(
    selector: 'heavy-loader',
    template: '<span>heavy loader #{{id}} on duty!</span>')
class HeavyLoaderComponent implements OnInit, OnDestroy {
  int id = nextId++;
  @Input() List<String> logs;

  // Mock todo: get 10,000 rows of data from the server
  ngOnInit() => _log(
      "heavy-loader $id initialized, loading 10,000 rows of data from the server");

  // Mock todo: clean-up
  ngOnDestroy() => _log("heavy-loader $id destroyed, cleaning up");

  _log(String msg) {
    logs.add(msg);
    _tick();
  }

  /// Triggers the next round of Angular change detection
  /// after one turn of the browser event loop
  /// ensuring display of msg added in onDestroy
  _tick() => new Future(() {});
}
// #enddocregion

File: public/docs/_examples/structural-directives/ts/app/heavy-loader.component.ts
// #docregion
import { Component, Input, OnDestroy, OnInit } from '@angular/core';

let nextId = 1;

@Component({
  selector: 'heavy-loader',
  template: '<span>heavy loader #{{id}} on duty!</span>'
})
export class HeavyLoaderComponent implements OnDestroy, OnInit {
  id = nextId++;
  @Input() logs: string[];

  ngOnInit() {
    // Mock todo: get 10,000 rows of data from the server
    this.log(`heavy-loader ${this.id} initialized,
      loading 10,000 rows of data from the server`);
  }

  ngOnDestroy() {
    // Mock todo: clean-up
    this.log(`heavy-loader ${this.id} destroyed, cleaning up`);
  }

  private log(msg: string) {
    this.logs.push(msg);
    this.tick();
  }

  // Triggers the next round of Angular change detection
  // after one turn of the browser event loop
  // ensuring display of msg added in onDestroy
  private tick() { setTimeout(() => { }, 0); }
}
// #enddocregion
